### PR TITLE
Add react datagrid

### PR DESCRIPTION
**Pull request by danrot**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | ---
| Related issues/PRs | ---
| License | MIT
| Documentation PR | ---

#### What's in this PR?

This PR adds the datagrid in react.

#### Why?

Because it is one of our most important components :-)

#### To Do

- [x] Write tests for Datagrid
- [x] Write tests for DatagridStore
- [x] Write tests for TableAdapter
- [x] Write tests for MetadataStore
- [ ] Think about how to load schema/urls from server
- [ ] Set sensible title in TableAdapter
- [x] Implement loading correctly
- [x] Add documentation for List component
- [x] Add documentation for Datagrid component

### CODE REVIEW

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Pagination/Pagination.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    render() {
+        const previousClass = classNames({
+            [paginationStyles.previous]: true,
+            [paginationStyles.enabled]: this.hasPreviousPage(),
+        });
+        const nextClass = classNames({
+            [paginationStyles.next]: true,
+            [paginationStyles.enabled]: this.hasNextPage(),
+        });
+        // TODO load translations in here?
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Pagination/Pagination.js]: This needs to be clarified before merge.

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Application/Application.js [hunk 2]
```diff
@@ -15,15 +15,18 @@ type Props = {
...
                 <Toolbar />
                 <main className={applicationStyles.main}>
-                    {this.props.router.currentRoute &&
+                    {router.currentRoute &&
                         <ViewRenderer
                             key={this.props.router.currentRoute.name}
                             name={this.props.router.currentRoute.view}
-                            parameters={this.props.router.currentParameters} />
+                            parameters={this.props.router.currentParameters}
+                            router={router} />
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/Application/Application.js]: This made me ask the question if we should omit the `parameters` prop above as well. Why should we pass it, if the `View` could get all information it needs from the `Router` directly? That would also mean that the `View`s have a very clear interface - only the router is passed.

Comment (danrot) [reply src/Sulu/Bundle/AdminBundle/Resources/js/containers/Application/Application.js]: On the other handside it is nicer to write `Views` with proper props being passed for a developer. But then it's the question how to pass the query parameters? Merging the query parameters might be convenient in some cases, but might allow to override parameters via query parameters even if they are not supposed to be used like that.

Comment (danrot) [reply src/Sulu/Bundle/AdminBundle/Resources/js/containers/Application/Application.js]: I've change that implementation for now in the other PR, which should be merge before this one: https://github.com/sulu/sulu/pull/3507/commits/b5574dd523ee0413a16a2dbd453439f3527fd0a1

We can discuss that, and revert the changes if we decide to handle this in a different way.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Pagination/Pagination.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        const nextClass = classNames({
+            [paginationStyles.next]: true,
+            [paginationStyles.enabled]: this.hasNextPage(),
+        });
+        const {current, total} = this.props;
+
+        return (
+            <nav className={paginationStyles.pagination}>
+                <span className={paginationStyles.display}>
+                    {translate('sulu_admin.page')}: {current} {translate('sulu_admin.of')} {total}
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Pagination/Pagination.js]: Not sure if we should build the translated part like this, or include https://github.com/yahoo/intl-messageformat now.
But the problem that would still persist is that the current variable should be in a textbox later, and I don't know how that should work in combination with translations. 😕

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/stores/DatagridStore.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
+// @flow
+import {action, autorun, observable} from 'mobx';
+import Requester from '../../../services/Requester';
+import metadataStore from './MetadataStore';
+
+export default class DatagridStore {
```
Comment (chirimoya) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/stores/DatagridStore.js]: `ResourceStore` ???

File: src/Sulu/Bundle/AdminBundle/Resources/js/views/List/List.js [hunk 5]
```diff
@@ -1,129 +1,53 @@
...
-    };
+    datagridStore: DatagridStore;
 
     componentWillMount() {
-        this.props.router.bindQuery('page', this.page, '1');
+        const router = this.props.router;
+        const {
+            route: {
+                options: {
+                    resourceKey,
```
Comment (chirimoya) [inline src/Sulu/Bundle/AdminBundle/Resources/js/views/List/List.js]: what if `resourceKey` doesn't exist in the route options? error handling?

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/stores/DatagridStore.js (hunk 4 above)
Comment (danrot) [reply src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/stores/DatagridStore.js]: But should it still be placed in the same folder? And when we are talking about that, should we still introduce some kind of `ResourceRequester`? Or do we wait with that until it becomes necessary? Might be already the case in the next PR, when try to delete elments from the datagrid.

File: src/Sulu/Bundle/AdminBundle/Resources/js/views/List/List.js (hunk 5 above)
Comment (danrot) [reply src/Sulu/Bundle/AdminBundle/Resources/js/views/List/List.js]: Good catch, I think I am throwing an error in that case. But I think that should also be handled somehow in our testing infrastructure... We need something similar as the behat tests now (but please no behat...), which just try to call all lists, and see if they are not failing.

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/stores/DatagridStore.js (hunk 4 above)
Comment (danrot) [reply src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/stores/DatagridStore.js]: When thinking about it... This store is already containing datagrid specific information, like the page. So should we simply rename it, and leave it where it is? And other components are not using it?

Comment (chirimoya) [reply src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/stores/DatagridStore.js]: keep `DatagridStore` for now and maybe move some parts into a `ResourceRequester `

File: src/Sulu/Bundle/AdminBundle/Resources/js/views/Form/Form.js [hunk 6]
```diff
@@ -39,7 +38,10 @@ class Form extends React.PureComponent<ViewProps> {
 export default withToolbar(Form, function() {
     return {
         backButton: {
```
Comment (chirimoya) [inline src/Sulu/Bundle/AdminBundle/Resources/js/views/Form/Form.js]: Disable backButton if no `router.route.options.backRoute` is available?

Comment (chirimoya) [reply src/Sulu/Bundle/AdminBundle/Resources/js/views/Form/Form.js]: don't add the button at all ;-)

**Pull request merged**